### PR TITLE
Enable persistence unit post processors configurable in EntityManager…

### PR DESCRIPTION
**Pull request by JohnNiang**
> Enable persistence unit post processors configurable in EntityManagerFactoryBuilder

This PR enables persistence unit post processors configurable by implementing [EntityManagerFactoryBuilderCustomizer.java](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/orm/jpa/EntityManagerFactoryBuilderCustomizer.java). It works only for [org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean#internalPersistenceUnitManager](https://github.com/spring-projects/spring-framework/blob/8c9c59eba0bf1d8efd01d4130e46f158c5bf9df0/spring-orm/src/main/java/org/springframework/orm/jpa/LocalContainerEntityManagerFactoryBean.java#L96).

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@JohnNiang Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=25443)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=25443) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@JohnNiang Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=25443)!

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java [hunk 1]
```diff
@@ -104,18 +107,34 @@ public void setBootstrapExecutor(AsyncTaskExecutor bootstrapExecutor) {
...
+		this.persistenceUnitPostProcessors = persistenceUnitPostProcessors;
+	}
+
 	/**
 	 * A fluent builder for a LocalContainerEntityManagerFactoryBean.
 	 */
 	public final class Builder {
 
-		private DataSource dataSource;
+		private final DataSource dataSource;
```
Comment (snicoll) [inline spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java:131]: Thanks but we prefer that such unrelated change be done in an isolated commit. Can you please revert that?

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java [hunk 2]
```diff
@@ -104,18 +107,34 @@ public void setBootstrapExecutor(AsyncTaskExecutor bootstrapExecutor) {
...
 
 		private String[] packagesToScan;
 
 		private String persistenceUnit;
 
-		private Map<String, Object> properties = new HashMap<>();
+		private final Map<String, Object> properties = new HashMap<>();
```
Comment (snicoll) [inline spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java:137]: Thanks but we prefer that such unrelated change be done in an isolated commit. Can you please revert that?

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java [hunk 3]
```diff
@@ -104,18 +107,34 @@ public void setBootstrapExecutor(AsyncTaskExecutor bootstrapExecutor) {
...
 	}
 
+	/**
+	 * Set the PersistenceUnitPostProcessors to be applied to the PersistenceUnitInfo used
+	 * for creating this EntityManagerFactory.
+	 * <p>
+	 * Such post-processors can, for example, register further entity classes and jar
+	 * files, in addition to the metadata read from {@code persistence.xml}.
+	 * <p>
+	 * <b>NOTE: Only applied if no external PersistenceUnitManager specified.</b>
```
Comment (snicoll) [inline spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java:117]: You wrote that in the PR description and I am not sure I understood what you're trying to convey here. In the context of this particular class, does it matter?

**Review by snicoll — COMMENTED** [review]
Thank you for the PR. Could you please provide a bit more context that led you to create this PR? 

We can't consider merging this until the code that is changed is exercised by unit tests. Can you please add that as well?

**Comment by JohnNiang** [comment]
@snicoll 

If the changes are applied, so that we could customize `MutablePersistenceUnitInfo` as follow:

```java
@Bean
EntityManagerFactoryBuilderCustomizer entityManagerFactoryBuilderCustomizer() {
	return builder -> {
		builder.setPersistenceUnitPostProcessors(pui -> {
			pui.addManagedClassName("customized.attribute.converter.class.name");
                        // other operations on MutablePersistenceUnitInfo
                        pui...
		});
	};
}
```

A scene is that we can register attribute converters into `org.hibernate.jpa.boot.spi.EntityManagerFactoryBuilder` programmatically. Because default fetching `managed class names` perform is scanning from class path only.

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java (hunk 3 above)
Comment (JohnNiang) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java:117]: I copied from `org.springframework.orm.jpa.persistenceunit.DefaultPersistenceUnitManager#setPersistenceUnitPostProcessors`

Comment (snicoll) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java:117]: There's no need to do that.

Comment (JohnNiang) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java:117]: So what should I do here? Remove or simplify theme.

Comment (snicoll) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java]: Simplify. You can look at other methods in this class for inspiration. More importantly, tests should be added.

Comment (JohnNiang) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java]: Okay, I'll try hard.

Comment (JohnNiang) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java]: There are not tests related `EntityManagerFactoryBuilder` present, could you provide any suggestions for incoming tests?

Comment (JohnNiang) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/orm/jpa/EntityManagerFactoryBuilder.java]: Maybe I should learn more about `org.springframework.boot.autoconfigure.orm.jpa.AbstractJpaAutoConfigurationTests`.

**Comment by snicoll** [comment]
@JohnNiang thank you for making your first contribution to Spring Boot.
